src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -389,43 +421,48 @@ protected virtual void Dispose(bool finalize)
 
         #region Host API
 
+        private static Solution CheckAndAddProject(Solution newSolution, ProjectInfo project)
+        {
+            CheckProjectIsNotInSolution(newSolution, project.Id);
+            return newSolution.AddProject(project);
+        }
+
         /// <summary>
         /// Call this method to respond to a solution being opened in the host environment.
         /// </summary>
         protected internal void OnSolutionAdded(SolutionInfo solutionInfo)
         {
-            using (_serializationLock.DisposableWait())
-            {
-                var oldSolution = this.CurrentSolution;
-                var solutionId = solutionInfo.Id;
+            this.SetCurrentSolution(
+                oldSolution =>
+                {
+                    CheckSolutionIsEmpty(oldSolution);
 
-                CheckSolutionIsEmpty();
-                this.SetCurrentSolution(this.CreateSolution(solutionInfo));
+                    var newSolution = this.CreateSolution(solutionInfo);
 
-                solutionInfo.Projects.Do(p => OnProjectAdded_NoLock(p, silent: true));
+                    foreach (var project in solutionInfo.Projects)
+                        newSolution = CheckAndAddProject(newSolution, project);
 
-                var newSolution = this.CurrentSolution;
-                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.SolutionAdded, oldSolution, newSolution);
-            }
+                    return newSolution;
+                }, WorkspaceChangeKind.SolutionAdded);
```

Comment:
an example of a simple transformation (that doesn't mutate any other workspace state).  It only passes in the trnasformation (and the change-event info).  It doesn't need to do any onBefore/onAfter stuff.